### PR TITLE
Integrate AI in testsuite

### PR DESCRIPTION
**Pull request by edoardob90**
Main goal is to improve the feedback given when a test fails. This PR will supersede #68, integrating Copilot-level explanations of any error traceback obtained when running the tests.

Here's an example of the preliminary results from the `intro.ipynb` notebook:

image

### Notes

- **Minimum requirement** is a valid API key from OpenAI. ~~It should be "loaded" with the utility function `set_key()` inside the module `tutorial.enable_ai`~~
- The initial prompt collates the function's code (anything that belongs to a `solution_*` function) and the traceback returned by `pytest`. Given this context, the model sometimes provides a step-by-step solution (if the task is easy)
- Asking for an explanation triggers the API, hence the function call is cached if the query is the same (function's code + traceback). There's no way to call the API again without changing the solution code.

### TODO

- [x] Improve the prompt, making it less prone to provide a super-detailed solution
- [x] Better way of handling the API key secret
- [x] Improve response formatting (now it's just basic HTML)

### CODE REVIEW

**Comment by edoardob90** [comment]
Right now, to set the API key one must do:

```python
from tutorial.ai import config
API_KEY = "sk-*******" # your key
config(API_KEY)
```

Two more options, maybe useful for testing:
- `model`: change default model (default: `gpt-4o-mini`)
- `language`: the default language the model will use (default: English)

**Comment by edoardob90** [comment]
### TODO

- [ ] Add relevant `AIExplanation` objects when other errors occur (like `TEST_ERROR` as discussed in #253)
- [ ] Add a customizable config from a file saved on disk (to avoid manual setting in every notebook)
- [ ] Enable a custom query (the main logic is already there)

**Comment by edoardob90** [comment]
Check out #257 to an alternative (better?) to the caching mechanism.

**Comment by edoardob90** [comment]
WIP to merge #260 first before double-checking that everything works as expected

**Comment by edoardob90** [comment]
PR #260 changed substantially the UI of the tests (including more pytest output added to the test result output) and the AI widgets. Please, give your feedback on whether there's some "mismatch" in the UI.

A couple of things I'm still considering to review:
- The "Reveal solution" box: it doesn't use a dropdown anymore. Maybe it was better before?
- The AI explanation steps: they still use `ipywidgets.Accordion`. Maybe a tab-view is better since it's more compact but organized?

**Comment by Snowwpanda** [comment]
Missing depencencies: python-dotenv, markdown2, openai, tenacity should be added to the environment list.

**Comment by edoardob90** [comment]
> Missing depencencies: python-dotenv, markdown2, openai, tenacity should be added to the environment list.

Done in 402cff0

**Comment by edoardob90** [comment]
> I am trying to test this with the `object-oriented` notebook. Here I provided a solution that is syntactically correct, but clearly wrong and I get no output at all:
> 
> image

Thanks, Despina, that was a very well-hidden bug.

In general, you can use `%%ipytest debug` and our cell magic will spit out any output from IPython, that we suppress by default.

**Review by Snowwpanda — CHANGES_REQUESTED** [review]
Nice work, some things i found i commented below, mainly missing packages. I may have a second look later.

File: tutorial/tests/testsuite/helpers.py [hunk 1]
```diff
@@ -128,62 +301,173 @@ def display_results(self) -> None:
...
+                <div class="solution-box">
+                    <div class="solution-code">
+                        <div id="solution-content">
+                            {Code(data=self.solution, language="python")._repr_html_()}
+                        </div>
+                        <div class="solution-overlay" id="solution-overlay">
+                            <button class="solution-button" onclick="
+                                document.getElementById('solution-overlay').style.opacity = '0';
+                                document.getElementById('solution-overlay').style.pointerEvents = 'none';
+                                this.style.display = 'none';
```
[the remark this answers is not in the document]
Comment (Snowwpanda) [reply tutorial/tests/testsuite/helpers.py:442]: This may actually be more convoluted, i'm fine if we ignore the problem for now

**Review by Snowwpanda — APPROVED** [review]

File: tutorial/tests/testsuite/testsuite.py [hunk 2]
```diff
@@ -298,5 +308,39 @@ def load_ipython_extension(ipython):
...
+    else:
+        ipython.openai_client = None
+        message = (
+            "<div style='background-color: #ffebee; border-radius: 5px; padding: 10px;'>"
+            "🚫 <strong style='color: red;'>OpenAI API key is undefined.</strong><br>"
+            "If you want to enable AI, store your key as <code>OPENAI_API_KEY='sk-****'</code> "
+            "in the file <code>openai.env</code>. "
+            "Check the file <code>openai.env.example</code> for guidance.</div>"
+        )
+        display(HTML(message))
```
[the remark this answers is not in the document]
Comment (edoardob90) [reply tutorial/tests/testsuite/testsuite.py]: Addressed with 0a60172

File: tutorial/tests/testsuite/testsuite.py [hunk 3]
```diff
@@ -298,5 +308,39 @@ def load_ipython_extension(ipython):
...
+            "✅ <strong>OpenAI client configured successfully.</strong></div>"
+        )
+        display(HTML(message))
+    else:
+        ipython.openai_client = None
+        message = (
+            "<div style='background-color: #ffebee; border-radius: 5px; padding: 10px;'>"
+            "🚫 <strong style='color: red;'>OpenAI API key is undefined.</strong><br>"
+            "If you want to enable AI, store your key as <code>OPENAI_API_KEY='sk-****'</code> "
+            "in the file <code>openai.env</code>. "
```
[the remark this answers is not in the document]
Comment (edoardob90) [reply tutorial/tests/testsuite/testsuite.py:360]: Done 1313e1e

**Comment by edoardob90** [comment]
@Snowwpanda @despadam If you could give a final check, then I guess we're good 🙏🏻

**Comment by despadam** [comment]
> PR #260 changed substantially the UI of the tests (including more pytest output added to the test result output) and the AI widgets. Please, give your feedback on whether there's some "mismatch" in the UI.
> 
> A couple of things I'm still considering to review:
> 
> * The "Reveal solution" box: it doesn't use a dropdown anymore. Maybe it was better before?
> * The AI explanation steps: they still use `ipywidgets.Accordion`. Maybe a tab-view is better since it's more compact but organized?

Also, I agree with both suggestions here:
1. Being able to reveal/hide the solution
2. Tabs instead of accordion for the AI explanation

However, none of these is a priority, so I am happy if we merge it as it is and update these at a later point.

**Comment by edoardob90** [comment]
> Thanks for the great work, @edoardob90 ! My only comment is that (with AI or without) if I press `reveal solution` once in any exercise, then when I try to do it again in another exercise, it will not be revealed. Is there maybe an internal counter or flag somewhere that handles this? Perhaps this issue is also related to @Snowwpanda 's comment above.

This is clearly a bug then, I need to check it carefully. I bet it's some JS or CSS. I'll do it after finishing some materials update otherwise we're going to be too late with that.

**Comment by edoardob90** [comment]
The issue was with the new styling of the "Reveal" action. I fixed it, but I think it's better to restore the old collapsible layout. I left the new styling as a reference for now. We might want to remove it if it's unnecessary.

File: tutorial/tests/testsuite/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (edoardob90) [reply tutorial/tests/testsuite/helpers.py:540]: See [comment](https://github.com/empa-scientific-it/python-tutorial/pull/248#issuecomment-2497525168).

**Review by despadam — APPROVED** [review]
Works perfectly fine for me now. I believe we can go ahead and merge 🚀

**Comment by edoardob90** [comment]
Great, let's go. If a last-minute bug comes up, we'll fix it.

**Pull request merged**